File: ieclass/exec1.py
import numpy as np
import matplotlib.pyplot as plt

# %%%%%%%%%%%%%%%%%%%%%%%%%%%
# % dados de simulacao
# % \dot{x}(t) = -2x(t)+bu(t)
# % x = (b/(s+2))u

# %%%%%%%%%%%%%%%%%%%%%%%%%%%
# Dados de simulação
b = 3
num = [b]
den = [1, -2]  # Ajuste o sinal do coeficiente '2' para '-2' para representar o sistema correto
N = 300  # Número de amostras
dt = 0.01
T = np.arange(0, N * dt, dt)
u = np.ones(N)
y = np.zeros(N)

# %%%%%%%%%%%%%%%%%%%%%%%%%%%
# Algoritmo de identificação de parâmetros
teta = np.zeros(N)
erro = np.zeros(N)
phi = np.zeros(N)
gama = 1000
ms = 10

# %%%%%%%%%%%%%%%%%%%%%%%%%%%
for k in range(N):
    y[k] = (b / (1 - np.exp(-2 * T[k])))  # Simulação da planta
    teta[k] = teta[k] + gama * dt * ((y[k] - teta[k] * phi[k]) / ms) * phi[k]
    phi[k] = u[k] - 2 * phi[k] * dt
    erro[k] = y[k] - teta[k] * phi[k]

# %%%%%%%%%%%%%%%%%%%%%%%%%%%
# Gráfico dos parâmetros identificados e do erro
plt.figure(figsize=(10, 6))
plt.plot(T, teta, 'b', linewidth=2, label='Parâmetro b')
plt.plot(T, erro, 'r', linewidth=2, label='Erro')
plt.legend()
plt.xlabel('Tempo')
plt.ylabel('Parâmetro Identificado / Erro')
plt.grid(True)
plt.show()

File: ClassExamples/AdaptativeControl/exec1.m
close all
clear all

%%%%%%%%%%%%%%%%%%%%%%%%%%%
% dados de simulacao
% \dot{x}(t) = -2x(t)+bu(t)
% x = (b/(s+2))u

%%%%%%%%%%%%%%%%%%%%%%%%%%%

b     = 3;
num   = b;
den   = [1 2];
ftx   = tf(num,den);
N     = 3;
dt    = 0.01;
T     = 0:dt:N;
u     = ones(1,N/dt+1); 
[Y,T] = lsim(ftx,u,T);

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% algoritmo de identificacao de parametros
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

teta(1)     = 0;
erro(1)     = 0;
ftx1        = tf(1,den);
[phi,T]     = lsim(ftx1,u,T);
gama        = 1000;
ms          = 10;

for k = 1:N/dt
    
    teta(k+1) = teta(k) + gama*dt*((Y(k) - teta(k)*phi(k))/ms)*phi(k);
    erro(k+1)   = Y(k) - teta(k)*phi(k);
end

figure(1)
plot(T,teta,'LineWidth',2)
hold on
plot(T,erro,'LineWidth',2)
legend('Parametro b','Erro')
xlabel('amostras')
ylabel('Parametro Identificado')
ax =gca;
ax.FontSize = 14;
hold off
saveas(1,'ex1_agosto_2021_a')

% Gráfico do erro
figure(2)
plot(T,erro,'linewidth',2)
xlabel('samples')
ylabel('erro')
ax

